### DISCUSSION TITLE
How to I read logs, how do I find the firmware version running on my CPC200-U2W Plus?

### DISCUSSION THREAD

**Discussion by Qhilm** [ludwig-v/wireless-carplay-dongle-reverse-engineering#190] [Q&A]
Hello, I'm trying to determine, as a first step, the firmware version I'm running. How can I best achieve this? I would also like to look at logs.

I checked the different ReadMes, FAQs and discussions here, but I am not any closer.

Probably so obvious no one bothered to write it down, but I can't figure it out. If someone could point me in the right directions, that would be much appreciated.

**Comment by Qhilm** [comment]
I found this in the HTML of the page served under "192.168.50.2", but somehow I doubt it's the firmware version:

```html
<script src="lang.js?v=20201210103216"></script>
```

**Comment by Qhilm** [comment]
I bought my CPC200-U2W Plus in February 2021, hence I should be on a firmware release ≤ 2021.03.06.

Unfortunately, when I plug a FAT32 formatted USB drive with the latest U2W_Update.img on it, it does not work. USB drive blinks a bit, then it stops and the LED on the carlinkit device remains solid red all the time. (it is not connected to a car on the USB-C side, it connected to my computer).

I should use the USB-A port for the USB dongle, right?

**Comment by Qhilm** [comment]
#### My goal
disable TKIP in the `hostapd.conf` file, as described [here](https://github.com/segfly/carlinkit-modding).

#### Summarizing open questions:
1. How do I find out the firmware my CPC200-U2W Plus bought end of Feb. 2021 could be running?
2. What are the rules for switching to a customer firmware? Should I pick the _BASED img file of exactly the stock firmware I'm currently running?
3. Why are custom firmwares all named `U2W_Update.img` even the old ones, and there is no `U2W_ZYD_Update.img` file? The Readme in the firmware folder states, that `U2W_ZYD_Update.img` is the required name "[if you are on 2020.08.27/2021.03.10](https://github.com/ludwig-v/wireless-carplay-dongle-reverse-engineering/tree/master/Firmware/U2W)")
4. Why would one need to compile, is it only to add own specific functionalities?
5. Should I upgrade the stock firmware first and then switch to custom?

Sorry if it's a lot of questions, but I'm struggling a bit to make sense of all readmes.

#### What I tried so far
* `2021.08.24_BASED/U2W_Update.img`
* `2021.03.06_BASED\U2W_Update.img`
* `U2W/_ZYD/2021.03.10.1430/U2W_Update.img`

None of these cause the solid red LED on my CPC200-U2W Plus to do anything. Each time it's the same, the USB stick blinks for a couple of seconds, then it stops and nothing happens anymore.

**Reply by j0wns** [reply]
BEWARE:  Read to the end or you may break stuff.

This request will return json containing your current version.
`http://192.168.50.2/cgi-bin/server.cgi?cmd=getversion`

{"version":"2020.08.27.1759","type":"IU","boxMac":"44:01:bb:XX:XX:XX","fileName":"U2W_ZYD_Update.img"}



I got my device in November of 2020.
Today I began in a similar place as you. I found a version that WOULD patch my system but now my device just shows me a QR Code on my vehicle's dash.
I'm in the middle of the mire segfly describes in the page you linked.
If your device currently works (provides wireless carplay) and you want it to continue working, I suggest you stop while you're ahead.
If you want to forge ahead, go ahead and preorder the CH341a programmer so you don't have to wait if something breaks.

The filename that upgraded my device from 2020.08.27.1759 to 2021.03.10.1430 is 'U2W_ZYD_Update.img'

**Reply by Qhilm** [reply]
thanks, but I already bricked it. I get the wifi, but the car does not recognize it anymore, just as described [here](https://github.com/segfly/carlinkit-modding#:~:text=car's%20infotainment%20unit%20no%20longer%20recognized%20the%20device%20as%20a%20CarPlay%20unit).

I have a raspberry pi 3 which I intend to use with flashrom, but my soldering skills are limited, I really struggle with how tiny this is.

**Reply by j0wns** [reply]
I'm going the USB flash tool route with the C341A. Soldering seems like a great way to make my responsive yet non-functional device completely useless.

**Reply by Qhilm** [reply]
Actually, you are right, all I need is a "Programmer Testing Clip SOP8 SOP16" it seems, and I can connect it to the raspberry Pi 3.

I have never done this and [this (very talented) guy](https://github.com/segfly/carlinkit-modding#:~:text=car's%20infotainment%20unit%20no%20longer%20recognized%20the%20device%20as%20a%20CarPlay%20unit) did it with soldering and I thought, sure, let me try that too.

**Reply by j0wns** [reply]
Make sure you take a backup of your device before going further.

I'm able to flash to @bronken [dump] firmware but can't activate.
This is my activation state now if not connected to a vehicle:
{"uuid":"5988f5252409b1d7adf828816a307e56","old_uuid":"5988f541142b89d7798d4bf009515697", "manufacturer":"GuanSheng","isActivated":1,"code":"ENl0iOoCrCeuvbekRGemipCXDnxYa6YA","boxType":"PU","carManufacturer":"Unknown","carModel":"Unknown","carOemName":"Unknown","carResolution":"0x0","productType":"U2W"}
It will connect to the vehicle and appear in the dash but then I get an activation screen.

Trying to figure out that to do next
<img width="826" alt="Screenshot 2024-11-21 at 5 45 20 PM" src="https://github.com/user-attachments/assets/0d2c3c57-0fbb-4b9e-ae73-6c6cd0320c7b">
<img width="963" alt="Screenshot 2024-11-21 at 5 45 00 PM" src="https://github.com/user-attachments/assets/3227dd6d-ca33-4ef3-8e7a-f351e8a0d766">
<img width="966" alt="Screenshot 2024-11-21 at 5 44 53 PM" src="https://github.com/user-attachments/assets/dc3f319d-b85e-42b7-a6ba-a397abc08581">

I think I need to get back to 2020.08.27.1759. When my device worked, its filename was U2W_ZYD_Update.img

**Reply by Qhilm** [reply]
Still according to [this](https://github.com/segfly/carlinkit-modding#:~:text=/etc/uuid_sign), the important bit for the activation is the `/etc/uuid_sign` file, which I already backed up.

But I'll take your advice about backing up the entire ROM again.

Something like `flashrom -p internal -r backup.rom -o backuplog.txt` should do the trick, we'll see. I'm waiting for my SOIC clip to arrive.

**Reply by j0wns** [reply]
> Still according to [this](https://github.com/segfly/carlinkit-modding#:~:text=/etc/uuid_sign), the important bit for the activation is the `/etc/uuid_sign` file, which I already backed up.
> 
> But I'll take your advice about backing up the entire ROM again.
> 
> Something like `flashrom -p internal -r backup.rom -o backuplog.txt` should do the trick, we'll see. I'm waiting for my SOIC clip to arrive.

Precisely how did you go about extracting the /etc/uuid_sign?

**Reply by Qhilm** [reply]
I installed dropbear using a [custom firmware](https://github.com/ludwig-v/wireless-carplay-dongle-reverse-engineering/tree/master/Custom_Firmware/Scripts/Dropbear), then ssh to the box, then `cat /etc/uuid_sign`.

**Reply by Zubastic** [reply]
@Qhilm dropbear could be only on custom firmware? I having different firmware, but want to extract it.

**Reply by Qhilm** [reply]
No idea, sorry.

**Reply by Zubastic** [reply]
> I think I need to get back to 2020.08.27.1759. When my device worked, its filename was U2W_ZYD_Update.img

Well, I having same board and update file name is `U2W_AUTOKIT_Update.img`.

After flashing 2020.12.25.1732_ROLLBACK nothing happend with any firmware before 2021.08.24.1817. Only this firmware accepted by dongle.

So, how to use rollback firmware if dongle refuses to use it?

Today I will burn flash dump, but it is interesting how to do that.

**Reply by ludwig-v** [reply]
@Zubastic Stop spamming the repo in every possible discussion or issue, it is extremely annoying. You have all the documentation available inside the repo, don’t use it as your brain thinking.
It is around 2021.07 or 2021.08 that they hardened the kernel to block strace so wifi drivers compiled later are not working on previous kernels (due to this recompilation), it is most likely the reason why the rollback is failing especially if you don’t have a Realtek RTL8822S !

**Reply by Zubastic** [reply]
@ludwig-v ok, but installing custom fw is not clear. I waste about 5 hours to bruteforce process of rollback. Initial firmware of that dongle is 2021.03.06, but I can’t restore it back. I waste this fw, so it is impossible to get it back, as I know. When it starts process green led (at repo info that it is blue) blinking, but when use old fw nothing signals that process started. It is confusing.

So my path was:
1. Initial 2021.03.06
2. Rollback 2020.12.25
3. Any fw is refused to flash until 2021.08.24.

How it is impossible to rollback if you haven’t RTL8822S right?

**Reply by ludwig-v** [reply]
Take a picture of your board.

Dealing with Carlinkit firmwares is not clear itself, reminder: it is not an image, but a tarball archive, so it is not always removing or adding all needed files depending on the path you took (original firmware to firmware upgrades / downgrades).

Back in 2020 a lot of people had their device bricked due to this poorly design firmware upgrades / downgrades.

**Reply by Zubastic** [reply]
Here it is:
![camphoto_1804928587](https://github.com/user-attachments/assets/5732ebb1-6d1a-4669-b047-b14c1160c834)

![image](https://github.com/user-attachments/assets/15f4b8cb-fa35-4dde-8ddf-47172b5ee07d)

![image](https://github.com/user-attachments/assets/10219a65-4ea7-42a1-b7eb-401cac9c7c40)

**Reply by ludwig-v** [reply]
Rollback firmware probably does not work for you because you have LGX8354S (Broadcom 4354), drivers started to be included inside 2021.03.06
Try ROLLBACK firmwares from U2W (normal) and rename files to U2W_AUTOKIT

**Reply by ludwig-v** [reply]
If you managed to flash 2020.12.25, why do you say that firmwares are refusing to flash ? If it is only because the device appears "dead" it is just that you must install your wifi drivers manually, if the device sticks with 2020.12.25 then yes, okey
If you can flash Custom Firmware 2021.08.24 as you said, you can put the Downgrade script (U2W.sh inside USB key) and flash 2021.03.06 **as written inside README** ..

**Reply by Zubastic** [reply]
> If you managed to flash 2020.12.25, why do you say that firmwares are refusing to flash ?

Rollback fw flashed normally changing version to 2020.12.25 but too quickly (green led not flashing at all, but version changed). I don’t know is it ok or not.

> If you can flash Custom Firmware 2021.08.24 as you said, you can put the Downgrade script (U2W.sh inside USB key) and flash 2021.03.06 as written inside README .

No, I could only flash original fw without any custom. So I in a loop: downgrade to 20.12.25 and upgrade to 2021.08.24.

I could after downgrade script use custom U2W.sh or not?

**Reply by Zubastic** [reply]
It should works like this:
1. Downgrade using rollback fw 2020.12.25
2. Install original fw 2021.02.05.2216
3. Install custom fw 2021.03.06

 Right sequence?

**Reply by ludwig-v** [reply]
The method is correct, but ROLLBACK firmware is not flashed properly (again: probably due to Wifi drivers), you shouldn't be able to install 2021.08.24 Official/Stock if it was downgraded correctly. You have no choice: hardware method.

**Reply by Zubastic** [reply]
Ok, thanks. I will do it today.

You build any gdbserver binary like Henkru did?

**Comment by Qhilm** [comment]
Just to go back to my topic, as a way of an update, I think I bricked mine. WiFi still working but car doesn't recognise it anymore.

I have tried using a Pi3 to unbrick it, but flashrom insists that "No EEPROM/flash device found". I checked cabling, I swapped out all cables, didn't help.  had sometimes the red LED on on the board (when the chip was powered by the Pi3), sometimes the green one, lately no LED at all turns on, maybe something fried.

So I bought a new one (still a CPC200-U2W Plus, but running 2023.05.13.1107) and hence am still rocking a WiFi with a stupidly easy to guess static password, which I strongly dislike, but I can't solve that. I really don't understand how they "harden" their config (disabling strace as @ludwig-v mentioned above) but don't provide a way to change the [6th most used password](https://en.wikipedia.org/wiki/List_of_the_most_common_passwords) in the world.

I still have the old one, happy to try other things to reflash it if anyone has an idea.

**Reply by ludwig-v** [reply]
You can buy a cheap CH341A Programmer of Amazon and use the clip to flash using ASprogrammer instead of the Pi ;)
If your flash is dead you will be missing the uuid_sign and the device, even restored with a valid backup, will be blocked in activation mode

@Henkru did work on this password issue: https://github.com/Henkru/cplay2air-wifi-passphrase-patch

**Reply by Qhilm** [reply]
>You can buy a cheap CH341A Programmer of Amazon and use the clip to flash using ASprogrammer instead of the Pi

Thanks, I had a Pi3 lying around, so I used it. The flash is not dead, wifi works, I also already backed up uuid_sign. But unless flashrom can detect it, there's not much I can do.

>@Henkru did work on this password issue: https://github.com/Henkru/cplay2air-wifi-passphrase-patch

Yes, I saw that, I still need to be able to downgrade as the requirements are:

>A CPLAY2air/Carlinkit dongle running ludwig-v's custom firmware. At the moment, the only supported version is 2021.03.06.

It's not taking 2020.11.10.2138_ROLLBACK though, I tried it, the gree LED does not turn on when the USB is plugged in. Actually, even my first carlinkit (the one which is now probably bricked) did not take 2020.11.10.2138_ROLLBACK even though it was taking all other firmwares.

**Reply by ludwig-v** [reply]
As far as I remember, I had more success with 2020.08.04.1602_ROLLBACK (From normal U2W), 2023.05.13 = AUTOKIT so you have to rename the file to U2W_AUTOKIT_Update.img
